### PR TITLE
[SPARK-23015][WINDOWS] Fix bug in Windows where starting multiple Spark instances within the same second causes a failure

### PR DESCRIPTION
**Pull request by Rafnel**
### What changes were proposed in this pull request?
**Problem**
If you attempt to start multiple Spark instances within a second there is a high likelihood that this spark-class-launcher-output file will have the same name for multiple instances, causing the Spark launcher to fail. The error will look something like this:

```
WARNING - The process cannot access the file because it is being used by another process.
WARNING - The system cannot find the file C:\Users\CURRENTUSER\AppData\Local\Temp\spark-class-launcher-output-21229.txt.
WARNING - The process cannot access the file because it is being used by another process.
```

Windows' %RANDOM% is seeded with 1-second granularity. We often start ~20 instances at the same time daily in Windows and encounter this bug on a weekly basis

**Proposed Fix**
Instead of relying on %RANDOM% which has poor granularity, use Powershell to generate a GUID and append that to the end of the temp file name. We have been using this in production for around 2-3 months and have never encountered this bug since.

### Why are the changes needed?
My team runs Spark on Windows and we boot up 20+ instances within a few seconds on a daily basis. We encountered this bug weekly and have taken steps to mitigate it without changing the Spark source code like adding a random sleep between 1-300 seconds before starting Spark. Even with a random sleep, 20+ instances have a likelihood of sleeping a similar amount of time and starting at the same time. Also, relying on a random sleep before starting Spark is clunky, unreliable, and not a deterministic way to avoid this issue.

Eventually our team went ahead and edited the code in this .cmd file with this fix. I figured I should make a pull request for this as well.

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
You can pretty reliably recreate this bug by submitting 30 Spark jobs in Windows using spark-submit. Eventually the Spark launcher will overlap with another Spark launcher and fail.

You can pull my fixed spark-class2.cmd and try this again and there should be no incidence of this bug.

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

**Comment by Rafnel** [comment]
I enabled github actions on my forked repo but it's not clear to me how to re-kick the failed build now that I've done that. Perhaps a maintainer will know how to do this. I haven't used github's CI/CD before.

**Comment by Rafnel** [comment]
I also wanted to note someone attempted to fix this issue with this PR in 2019: https://github.com/apache/spark/pull/25076 and this did get merged, but does not appear to fix the issue. We are using Spark 3.2.1 which does have spark-25076 in the spark-class2.cmd but we encounter this bug all the same.

**Review by HyukjinKwon — COMMENTED** [review]
The change itself seems fine .. would be great if a committer can test this patch on Windows before merging it (since we don't have the CI for this - we do tests SparkR on Windows but not for others)

**Comment by Rafnel** [comment]
Wanted to chime in again and note that my team has been using this patched version of the spark-class2.cmd to launch our Spark instances for a couple weeks now and have not encountered the aforementioned bug anymore, or any other launch issues. 

Obviously this change doesn't //fix// the underlying issue, as you could still launch a Spark instance within the same 10ms period and encounter this bug, but it's much less likely than launching 2 Spark instances within the same second. My team launches around 20 instances in Windows within a couple seconds of each other each day and were seeing this issue almost daily until this patch.

**Comment by HyukjinKwon** [comment]
Let me ask the dev mailing list and see if we can have others test this patch

**Comment by HyukjinKwon** [comment]
Sent https://lists.apache.org/thread/9y0xrxo08cv3nt27wt40jrpncgzz1kt1

**Comment by LuciferYang** [comment]
friendly ping @panbingkun , if convenient, such as through offline communication, please help to verify this patch on Windows. Thanks ~

**Comment by panbingkun** [comment]
> friendly ping @panbingkun , if convenient, such as through offline communication, please help to verify this patch on Windows. Thanks ~

Okay, I'll verify it a little later.

**Comment by panbingkun** [comment]
I'm very sorry, I lost it. I will verify it on a Windows machine this weekend.

**Comment by LuciferYang** [comment]
remove the `Stale` tag

**Comment by Rafnel** [comment]
Hello all, I'm going to do one final edit on this PR. We found that `set LAUNCHER_OUTPUT=%temp%\spark-class-launcher-output-%RANDOM%%TIME::=0%.txt` still resulted in the occasional race condition error, as we launch tens of Spark instances at the same time each day on Windows. `%RANDOM%%TIME::=0%` still has a small chance of overlapping. Our team found that using GUIDs at the end of the temp file name completely fixed this bug and removed the need for %RANDOM% and %TIME%, entirely.

**Comment by Rafnel** [comment]
Alright, that should be my final edit. Like I said, we've had no issues since applying this fix a few months ago and we have since increased the amount of concurrent Spark instances that we launch on Windows.

File: bin/spark-class2.cmd
Comment (panbingkun) [inline bin/spark-class2.cmd:67]: If we use the `GUID` mechanism, do we still need this logic? 
```
:gen
...
if exist %LAUNCHER_OUTPUT% goto :gen
```

Comment (panbingkun) [reply bin/spark-class2.cmd:67]: In addition, the current logic strongly relies on `Powershell`.  I am worried about that the lack of `Powershell` on `older` versions of Windows will cause launch Spark failures. 
Can we `eliminate` this strong dependence?

**Comment by panbingkun** [comment]
This can indeed generate `a unique output file name` (I added some `print logs`), as follows:
image

image

And spark-shell starts normally.

File: bin/spark-class2.cmd
Comment (Rafnel) [reply bin/spark-class2.cmd:67]: > If we use the `GUID` mechanism, do we still need this logic?
> 
> ```
> :gen
> ...
> if exist %LAUNCHER_OUTPUT% goto :gen
> ```

You are correct, we shouldn't need this anymore. I can remove it.

Comment (Rafnel) [reply bin/spark-class2.cmd:67]: > In addition, the current logic strongly relies on `Powershell`. I am worried about that the lack of `Powershell` on `older` versions of Windows will cause launch Spark failures. Can we `eliminate` this strong dependence?

As for this concern, I think this is the most lightweight way to generate a GUID in a .cmd script (without installing any dependencies). Powershell is available in Windows 7+ and Windows Server 2008R2+. I would be surprised if there were users running Spark on Windows XP/Vista or Server versions older than 2008. In its current state, you can't reliably instantiate multiple Spark instances on Windows, so I think it's a worthwhile tradeoff.

Comment (Rafnel) [reply bin/spark-class2.cmd:66]: Updated to remove below goto statement.
```
:gen
...
if exist %LAUNCHER_OUTPUT% goto :gen
```

Comment (panbingkun) [reply bin/spark-class2.cmd:66]: > > In addition, the current logic strongly relies on `Powershell`. I am worried about that the lack of `Powershell` on `older` versions of Windows will cause launch Spark failures. Can we `eliminate` this strong dependence?
> 
> As for this concern, I think this is the most lightweight way to generate a GUID in a .cmd script (without installing any dependencies). Powershell is available in Windows 7+ and Windows Server 2008R2+. I would be surprised if there were users running Spark on Windows XP/Vista or Server versions older than 2008. In its current state, you can't reliably instantiate multiple Spark instances on Windows, so I think it's a worthwhile tradeoff.

Perhaps we can first check whether `Powershell` has been installed.  If not, use the original logic (until we find a simpler way to generate GUID), and if it is installed, use the logic based on `Powershell` to generate GUID?

Or prompt to install `Powershell` if it is not installed?

Comment (Rafnel) [reply bin/spark-class2.cmd]: Updated to add a simple check for powershell.exe and if not exists, fall back to using the Windows %RANDOM% which is the current behavior of this script.

**Comment by Rafnel** [comment]
Bumping this PR - any other concerns?

**Comment by panbingkun** [comment]
I think there is no problem with the overall logic, but I am a little worried about the `UUID generation`.
I suggest using `python` to generate `UUID`, because in our native logic, we have used `python` to generate `SPARK_HOME`,
This is to eliminate dependence on the outside, WDYT? @HyukjinKwon @LuciferYang 
https://github.com/apache/spark/blob/da92293f9ce0be1ac283c4a5d769af550abf7031/bin/find-spark-home.cmd#L21-L59

**Comment by HyukjinKwon** [comment]
problem is that python is not bundled with Windows IIRC ..

**Review by HyukjinKwon — APPROVED** [review]
I am fine with this given that it was tested properly.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**
